# November Analysis

In [75]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import warnings
warnings.filterwarnings("ignore")

First all the products are aggregated per category, taking as price of the category the mean of the log(price_changes) 

In [76]:
novembre = pd.read_csv('dataset/november.csv', index_col=0)

In [77]:
pdpdnovember['comp_date'] = pd.to_datetime(novembre['comp_date'])

In [78]:
ge=pdpdnovember.groupby(['comp_date','seller_id','coded_cat1'])['price_change_log'].mean().reset_index()
ge['comp_date'] = pd.to_datetime(ge['comp_date'])

pivot={}
for i in pdpdnovember['coded_cat1'].unique():
    pivot[i] = ge[ge['coded_cat1']==i].pivot(index='comp_date', columns='seller_id', values='price_change_log')

for i in pdpdnovember['coded_cat1'].unique():
    globals()['pivot{}'.format(i)] = pivot[i]

Dropping all the seller that have all null values for a specific category

In [79]:
c=0
for i in pdpdnovember['coded_cat1'].unique():
    for j in globals()['pivot{}'.format(i)].columns:
        if np.isnan(globals()['pivot{}'.format(i)][j]).all() == True or (globals()['pivot{}'.format(i)][j]).sum() == 0:
            print(f'for the seller {j} the category {i} has all null values or zeros, so we drop it')
            globals()['pivot{}'.format(i)].drop(j, inplace=True, axis=1)
        else:
            c += 1

if c == 93:
    print('None of the category has all null values for each of sellers')

None of the category has all null values for each of sellers


In [80]:
#check for stationarity and VAR model
from statsmodels.tsa.stattools import adfuller
from statsmodels.tsa.vector_ar.vecm import coint_johansen
from statsmodels.tsa.vector_ar.var_model import VAR

In [81]:
#fill na in all the pivot ffilled
for i in pdpdnovember['coded_cat1'].unique():
    globals()['pivot{}'.format(i)] = globals()['pivot{}'.format(i)].fillna(method='ffill')
    globals()['pivot{}'.format(i)] = globals()['pivot{}'.format(i)].fillna(method='bfill')
    globals()['pivot{}'.format(i)] = globals()['pivot{}'.format(i)].fillna(0)

## Pivot 1163

In [82]:
def adfuller_test(series, signif=0.05, name='', verbose=False):
    """Perform ADFuller to test for Stationarity of given series and print report"""
    r = adfuller(series, autolag='AIC')
    output = {'test_statistic':round(r[0], 4), 'pvalue':round(r[1], 4), 'n_lags':round(r[2], 4), 'n_obs':r[3]}
    p_value = output['pvalue'] 
    def adjust(val, length= 6): return str(val).ljust(length)

    # Print Summary
    print(f'    Augmented Dickey-Fuller Test on "{name}"', "\n   ", '-'*47)
    print(f' Null Hypothesis: Data has unit root. Non-Stationary.')
    print(f' Significance Level    = {signif}')
    print(f' Test Statistic        = {output["test_statistic"]}')
    print(f' No. Lags Chosen       = {output["n_lags"]}')

    for key,val in r[4].items():
        print(f' Critical value {adjust(key)} = {round(val, 3)}')

    if p_value <= signif:
        print(f" => P-Value = {p_value}. Rejecting Null Hypothesis.")
        print(f" => Series is Stationary.")
    else:
        print(f" => P-Value = {p_value}. Weak evidence to reject the Null Hypothesis.")
        print(f" => Series is Non-Stationary.")    

In [83]:
for name, column in pivot1163.iteritems():
    adfuller_test(column, name=column.name)
    print('\n')

    Augmented Dickey-Fuller Test on "23" 
    -----------------------------------------------
 Null Hypothesis: Data has unit root. Non-Stationary.
 Significance Level    = 0.05
 Test Statistic        = -5.3172
 No. Lags Chosen       = 0
 Critical value 1%     = -3.679
 Critical value 5%     = -2.968
 Critical value 10%    = -2.623
 => P-Value = 0.0. Rejecting Null Hypothesis.
 => Series is Stationary.


    Augmented Dickey-Fuller Test on "24" 
    -----------------------------------------------
 Null Hypothesis: Data has unit root. Non-Stationary.
 Significance Level    = 0.05
 Test Statistic        = -5.0648
 No. Lags Chosen       = 0
 Critical value 1%     = -3.679
 Critical value 5%     = -2.968
 Critical value 10%    = -2.623
 => P-Value = 0.0. Rejecting Null Hypothesis.
 => Series is Stationary.


    Augmented Dickey-Fuller Test on "26" 
    -----------------------------------------------
 Null Hypothesis: Data has unit root. Non-Stationary.
 Significance Level    = 0.05
 Test 

performing ADFuller Test to identify if the series are stationary or not

In [84]:
pivot1163 = pivot1163.diff().dropna()

In [85]:
for name, column in pivot1163.iteritems():
    adfuller_test(column, name=column.name)
    print('\n')

    Augmented Dickey-Fuller Test on "23" 
    -----------------------------------------------
 Null Hypothesis: Data has unit root. Non-Stationary.
 Significance Level    = 0.05
 Test Statistic        = -9.5675
 No. Lags Chosen       = 0
 Critical value 1%     = -3.689
 Critical value 5%     = -2.972
 Critical value 10%    = -2.625
 => P-Value = 0.0. Rejecting Null Hypothesis.
 => Series is Stationary.


    Augmented Dickey-Fuller Test on "24" 
    -----------------------------------------------
 Null Hypothesis: Data has unit root. Non-Stationary.
 Significance Level    = 0.05
 Test Statistic        = -5.8794
 No. Lags Chosen       = 1
 Critical value 1%     = -3.7
 Critical value 5%     = -2.976
 Critical value 10%    = -2.628
 => P-Value = 0.0. Rejecting Null Hypothesis.
 => Series is Stationary.


    Augmented Dickey-Fuller Test on "26" 
    -----------------------------------------------
 Null Hypothesis: Data has unit root. Non-Stationary.
 Significance Level    = 0.05
 Test St

# PIVOT 1375

In [86]:
 
for name, column in pivot1375.iteritems():
    adfuller_test(column, name=column.name)
    print('\n')

    Augmented Dickey-Fuller Test on "23" 
    -----------------------------------------------
 Null Hypothesis: Data has unit root. Non-Stationary.
 Significance Level    = 0.05
 Test Statistic        = -6.5483
 No. Lags Chosen       = 0
 Critical value 1%     = -3.679
 Critical value 5%     = -2.968
 Critical value 10%    = -2.623
 => P-Value = 0.0. Rejecting Null Hypothesis.
 => Series is Stationary.


    Augmented Dickey-Fuller Test on "24" 
    -----------------------------------------------
 Null Hypothesis: Data has unit root. Non-Stationary.
 Significance Level    = 0.05
 Test Statistic        = -4.8874
 No. Lags Chosen       = 0
 Critical value 1%     = -3.679
 Critical value 5%     = -2.968
 Critical value 10%    = -2.623
 => P-Value = 0.0. Rejecting Null Hypothesis.
 => Series is Stationary.


    Augmented Dickey-Fuller Test on "26" 
    -----------------------------------------------
 Null Hypothesis: Data has unit root. Non-Stationary.
 Significance Level    = 0.05
 Test 

In [87]:
pivot1375 = pivot1375.diff().dropna()

In [88]:
 
for name, column in pivot1375.iteritems():
    adfuller_test(column, name=column.name)
    print('\n')

    Augmented Dickey-Fuller Test on "23" 
    -----------------------------------------------
 Null Hypothesis: Data has unit root. Non-Stationary.
 Significance Level    = 0.05
 Test Statistic        = -7.4049
 No. Lags Chosen       = 1
 Critical value 1%     = -3.7
 Critical value 5%     = -2.976
 Critical value 10%    = -2.628
 => P-Value = 0.0. Rejecting Null Hypothesis.
 => Series is Stationary.


    Augmented Dickey-Fuller Test on "24" 
    -----------------------------------------------
 Null Hypothesis: Data has unit root. Non-Stationary.
 Significance Level    = 0.05
 Test Statistic        = -8.4854
 No. Lags Chosen       = 0
 Critical value 1%     = -3.689
 Critical value 5%     = -2.972
 Critical value 10%    = -2.625
 => P-Value = 0.0. Rejecting Null Hypothesis.
 => Series is Stationary.


    Augmented Dickey-Fuller Test on "26" 
    -----------------------------------------------
 Null Hypothesis: Data has unit root. Non-Stationary.
 Significance Level    = 0.05
 Test St

# PIVOT 2180

In [89]:
for name, column in pivot2180.iteritems():
    adfuller_test(column, name=column.name)
    print('\n')

    Augmented Dickey-Fuller Test on "23" 
    -----------------------------------------------
 Null Hypothesis: Data has unit root. Non-Stationary.
 Significance Level    = 0.05
 Test Statistic        = -5.0138
 No. Lags Chosen       = 0
 Critical value 1%     = -3.679
 Critical value 5%     = -2.968
 Critical value 10%    = -2.623
 => P-Value = 0.0. Rejecting Null Hypothesis.
 => Series is Stationary.


    Augmented Dickey-Fuller Test on "24" 
    -----------------------------------------------
 Null Hypothesis: Data has unit root. Non-Stationary.
 Significance Level    = 0.05
 Test Statistic        = -1.6326
 No. Lags Chosen       = 5
 Critical value 1%     = -3.738
 Critical value 5%     = -2.992
 Critical value 10%    = -2.636
 => P-Value = 0.4661. Weak evidence to reject the Null Hypothesis.
 => Series is Non-Stationary.


    Augmented Dickey-Fuller Test on "26" 
    -----------------------------------------------
 Null Hypothesis: Data has unit root. Non-Stationary.
 Significan

In [90]:
pivot2180 = pivot2180.diff().dropna()

In [91]:
for name, column in pivot2180.iteritems():
    adfuller_test(column, name=column.name)
    print('\n')

    Augmented Dickey-Fuller Test on "23" 
    -----------------------------------------------
 Null Hypothesis: Data has unit root. Non-Stationary.
 Significance Level    = 0.05
 Test Statistic        = -5.2915
 No. Lags Chosen       = 0
 Critical value 1%     = -3.689
 Critical value 5%     = -2.972
 Critical value 10%    = -2.625
 => P-Value = 0.0. Rejecting Null Hypothesis.
 => Series is Stationary.


    Augmented Dickey-Fuller Test on "24" 
    -----------------------------------------------
 Null Hypothesis: Data has unit root. Non-Stationary.
 Significance Level    = 0.05
 Test Statistic        = -6.7097
 No. Lags Chosen       = 1
 Critical value 1%     = -3.7
 Critical value 5%     = -2.976
 Critical value 10%    = -2.628
 => P-Value = 0.0. Rejecting Null Hypothesis.
 => Series is Stationary.


    Augmented Dickey-Fuller Test on "26" 
    -----------------------------------------------
 Null Hypothesis: Data has unit root. Non-Stationary.
 Significance Level    = 0.05
 Test St

# PIVOT 1676

In [92]:
for name, column in pivot1676.iteritems():
    adfuller_test(column, name=column.name)
    print('\n')

    Augmented Dickey-Fuller Test on "23" 
    -----------------------------------------------
 Null Hypothesis: Data has unit root. Non-Stationary.
 Significance Level    = 0.05
 Test Statistic        = -4.6091
 No. Lags Chosen       = 0
 Critical value 1%     = -3.679
 Critical value 5%     = -2.968
 Critical value 10%    = -2.623
 => P-Value = 0.0001. Rejecting Null Hypothesis.
 => Series is Stationary.


    Augmented Dickey-Fuller Test on "24" 
    -----------------------------------------------
 Null Hypothesis: Data has unit root. Non-Stationary.
 Significance Level    = 0.05
 Test Statistic        = -3.4145
 No. Lags Chosen       = 8
 Critical value 1%     = -3.788
 Critical value 5%     = -3.013
 Critical value 10%    = -2.646
 => P-Value = 0.0105. Rejecting Null Hypothesis.
 => Series is Stationary.


    Augmented Dickey-Fuller Test on "26" 
    -----------------------------------------------
 Null Hypothesis: Data has unit root. Non-Stationary.
 Significance Level    = 0.05


In [93]:
pivot1676 = pivot1676.diff().dropna()

In [94]:
for name, column in pivot1676.iteritems():
    adfuller_test(column, name=column.name)
    print('\n')

    Augmented Dickey-Fuller Test on "23" 
    -----------------------------------------------
 Null Hypothesis: Data has unit root. Non-Stationary.
 Significance Level    = 0.05
 Test Statistic        = -7.4692
 No. Lags Chosen       = 0
 Critical value 1%     = -3.689
 Critical value 5%     = -2.972
 Critical value 10%    = -2.625
 => P-Value = 0.0. Rejecting Null Hypothesis.
 => Series is Stationary.


    Augmented Dickey-Fuller Test on "24" 
    -----------------------------------------------
 Null Hypothesis: Data has unit root. Non-Stationary.
 Significance Level    = 0.05
 Test Statistic        = -5.8057
 No. Lags Chosen       = 1
 Critical value 1%     = -3.7
 Critical value 5%     = -2.976
 Critical value 10%    = -2.628
 => P-Value = 0.0. Rejecting Null Hypothesis.
 => Series is Stationary.


    Augmented Dickey-Fuller Test on "26" 
    -----------------------------------------------
 Null Hypothesis: Data has unit root. Non-Stationary.
 Significance Level    = 0.05
 Test St

# PIVOT 2259

In [95]:
pivot2259 = pivot2259.diff().dropna()

In [96]:
for name, column in pivot2259.iteritems():
    adfuller_test(column, name=column.name)
    print('\n')

    Augmented Dickey-Fuller Test on "23" 
    -----------------------------------------------
 Null Hypothesis: Data has unit root. Non-Stationary.
 Significance Level    = 0.05
 Test Statistic        = -8.5531
 No. Lags Chosen       = 0
 Critical value 1%     = -3.689
 Critical value 5%     = -2.972
 Critical value 10%    = -2.625
 => P-Value = 0.0. Rejecting Null Hypothesis.
 => Series is Stationary.


    Augmented Dickey-Fuller Test on "24" 
    -----------------------------------------------
 Null Hypothesis: Data has unit root. Non-Stationary.
 Significance Level    = 0.05
 Test Statistic        = -3.8286
 No. Lags Chosen       = 3
 Critical value 1%     = -3.724
 Critical value 5%     = -2.986
 Critical value 10%    = -2.633
 => P-Value = 0.0026. Rejecting Null Hypothesis.
 => Series is Stationary.


    Augmented Dickey-Fuller Test on "26" 
    -----------------------------------------------
 Null Hypothesis: Data has unit root. Non-Stationary.
 Significance Level    = 0.05
 Te

# PIVOT 885

In [97]:
pivot885 = pivot885.diff().dropna()

In [98]:
for name, column in pivot885.iteritems():
    adfuller_test(column, name=column.name)
    print('\n')

    Augmented Dickey-Fuller Test on "23" 
    -----------------------------------------------
 Null Hypothesis: Data has unit root. Non-Stationary.
 Significance Level    = 0.05
 Test Statistic        = -4.7209
 No. Lags Chosen       = 1
 Critical value 1%     = -3.7
 Critical value 5%     = -2.976
 Critical value 10%    = -2.628
 => P-Value = 0.0001. Rejecting Null Hypothesis.
 => Series is Stationary.


    Augmented Dickey-Fuller Test on "24" 
    -----------------------------------------------
 Null Hypothesis: Data has unit root. Non-Stationary.
 Significance Level    = 0.05
 Test Statistic        = -9.1918
 No. Lags Chosen       = 0
 Critical value 1%     = -3.689
 Critical value 5%     = -2.972
 Critical value 10%    = -2.625
 => P-Value = 0.0. Rejecting Null Hypothesis.
 => Series is Stationary.


    Augmented Dickey-Fuller Test on "26" 
    -----------------------------------------------
 Null Hypothesis: Data has unit root. Non-Stationary.
 Significance Level    = 0.05
 Test

# PIVOT 2880

In [99]:
for name, column in pivot2880.iteritems():
    adfuller_test(column, name=column.name)
    print('\n')

    Augmented Dickey-Fuller Test on "24" 
    -----------------------------------------------
 Null Hypothesis: Data has unit root. Non-Stationary.
 Significance Level    = 0.05
 Test Statistic        = -5.1281
 No. Lags Chosen       = 0
 Critical value 1%     = -3.679
 Critical value 5%     = -2.968
 Critical value 10%    = -2.623
 => P-Value = 0.0. Rejecting Null Hypothesis.
 => Series is Stationary.


    Augmented Dickey-Fuller Test on "26" 
    -----------------------------------------------
 Null Hypothesis: Data has unit root. Non-Stationary.
 Significance Level    = 0.05
 Test Statistic        = -2.7015
 No. Lags Chosen       = 5
 Critical value 1%     = -3.738
 Critical value 5%     = -2.992
 Critical value 10%    = -2.636
 => P-Value = 0.0738. Weak evidence to reject the Null Hypothesis.
 => Series is Non-Stationary.


    Augmented Dickey-Fuller Test on "41" 
    -----------------------------------------------
 Null Hypothesis: Data has unit root. Non-Stationary.
 Significan

In [100]:
pivot2880 = pivot2880.diff().dropna()

In [101]:
for name, column in pivot2880.iteritems():
    adfuller_test(column, name=column.name)
    print('\n')

    Augmented Dickey-Fuller Test on "24" 
    -----------------------------------------------
 Null Hypothesis: Data has unit root. Non-Stationary.
 Significance Level    = 0.05
 Test Statistic        = -5.9512
 No. Lags Chosen       = 1
 Critical value 1%     = -3.7
 Critical value 5%     = -2.976
 Critical value 10%    = -2.628
 => P-Value = 0.0. Rejecting Null Hypothesis.
 => Series is Stationary.


    Augmented Dickey-Fuller Test on "26" 
    -----------------------------------------------
 Null Hypothesis: Data has unit root. Non-Stationary.
 Significance Level    = 0.05
 Test Statistic        = -3.2299
 No. Lags Chosen       = 5
 Critical value 1%     = -3.753
 Critical value 5%     = -2.998
 Critical value 10%    = -2.639
 => P-Value = 0.0183. Rejecting Null Hypothesis.
 => Series is Stationary.


    Augmented Dickey-Fuller Test on "41" 
    -----------------------------------------------
 Null Hypothesis: Data has unit root. Non-Stationary.
 Significance Level    = 0.05
 Test

# PIVOT 1127

In [102]:
for name, column in pivot1127.iteritems():
    adfuller_test(column, name=column.name)
    print('\n')

    Augmented Dickey-Fuller Test on "23" 
    -----------------------------------------------
 Null Hypothesis: Data has unit root. Non-Stationary.
 Significance Level    = 0.05
 Test Statistic        = -5.2215
 No. Lags Chosen       = 0
 Critical value 1%     = -3.679
 Critical value 5%     = -2.968
 Critical value 10%    = -2.623
 => P-Value = 0.0. Rejecting Null Hypothesis.
 => Series is Stationary.


    Augmented Dickey-Fuller Test on "24" 
    -----------------------------------------------
 Null Hypothesis: Data has unit root. Non-Stationary.
 Significance Level    = 0.05
 Test Statistic        = -5.1106
 No. Lags Chosen       = 0
 Critical value 1%     = -3.679
 Critical value 5%     = -2.968
 Critical value 10%    = -2.623
 => P-Value = 0.0. Rejecting Null Hypothesis.
 => Series is Stationary.


    Augmented Dickey-Fuller Test on "26" 
    -----------------------------------------------
 Null Hypothesis: Data has unit root. Non-Stationary.
 Significance Level    = 0.05
 Test 

In [103]:
pivot1127 = pivot1127.diff().dropna()

In [104]:
for name, column in pivot1127.iteritems():
    adfuller_test(column, name=column.name)
    print('\n')

    Augmented Dickey-Fuller Test on "23" 
    -----------------------------------------------
 Null Hypothesis: Data has unit root. Non-Stationary.
 Significance Level    = 0.05
 Test Statistic        = -6.0
 No. Lags Chosen       = 1
 Critical value 1%     = -3.7
 Critical value 5%     = -2.976
 Critical value 10%    = -2.628
 => P-Value = 0.0. Rejecting Null Hypothesis.
 => Series is Stationary.


    Augmented Dickey-Fuller Test on "24" 
    -----------------------------------------------
 Null Hypothesis: Data has unit root. Non-Stationary.
 Significance Level    = 0.05
 Test Statistic        = -8.7704
 No. Lags Chosen       = 0
 Critical value 1%     = -3.689
 Critical value 5%     = -2.972
 Critical value 10%    = -2.625
 => P-Value = 0.0. Rejecting Null Hypothesis.
 => Series is Stationary.


    Augmented Dickey-Fuller Test on "26" 
    -----------------------------------------------
 Null Hypothesis: Data has unit root. Non-Stationary.
 Significance Level    = 0.05
 Test Stati

# PIVOT 1354

In [105]:
for name, column in pivot1354.iteritems():
    adfuller_test(column, name=column.name)
    print('\n')

    Augmented Dickey-Fuller Test on "23" 
    -----------------------------------------------
 Null Hypothesis: Data has unit root. Non-Stationary.
 Significance Level    = 0.05
 Test Statistic        = -0.8305
 No. Lags Chosen       = 5
 Critical value 1%     = -3.738
 Critical value 5%     = -2.992
 Critical value 10%    = -2.636
 => P-Value = 0.81. Weak evidence to reject the Null Hypothesis.
 => Series is Non-Stationary.


    Augmented Dickey-Fuller Test on "24" 
    -----------------------------------------------
 Null Hypothesis: Data has unit root. Non-Stationary.
 Significance Level    = 0.05
 Test Statistic        = -4.4187
 No. Lags Chosen       = 3
 Critical value 1%     = -3.711
 Critical value 5%     = -2.981
 Critical value 10%    = -2.63
 => P-Value = 0.0003. Rejecting Null Hypothesis.
 => Series is Stationary.


    Augmented Dickey-Fuller Test on "26" 
    -----------------------------------------------
 Null Hypothesis: Data has unit root. Non-Stationary.
 Significan

In [106]:
pivot1354 = pivot1354.diff().dropna()

In [107]:
for name, column in pivot1354.iteritems():
    adfuller_test(column, name=column.name)
    print('\n')

    Augmented Dickey-Fuller Test on "23" 
    -----------------------------------------------
 Null Hypothesis: Data has unit root. Non-Stationary.
 Significance Level    = 0.05
 Test Statistic        = -7.6157
 No. Lags Chosen       = 4
 Critical value 1%     = -3.738
 Critical value 5%     = -2.992
 Critical value 10%    = -2.636
 => P-Value = 0.0. Rejecting Null Hypothesis.
 => Series is Stationary.


    Augmented Dickey-Fuller Test on "24" 
    -----------------------------------------------
 Null Hypothesis: Data has unit root. Non-Stationary.
 Significance Level    = 0.05
 Test Statistic        = -4.0725
 No. Lags Chosen       = 4
 Critical value 1%     = -3.738
 Critical value 5%     = -2.992
 Critical value 10%    = -2.636
 => P-Value = 0.0011. Rejecting Null Hypothesis.
 => Series is Stationary.


    Augmented Dickey-Fuller Test on "26" 
    -----------------------------------------------
 Null Hypothesis: Data has unit root. Non-Stationary.
 Significance Level    = 0.05
 Te

# PIVOT 1617

In [108]:
for name, column in pivot1617.iteritems():
    adfuller_test(column, name=column.name)
    print('\n')

    Augmented Dickey-Fuller Test on "24" 
    -----------------------------------------------
 Null Hypothesis: Data has unit root. Non-Stationary.
 Significance Level    = 0.05
 Test Statistic        = -5.6628
 No. Lags Chosen       = 0
 Critical value 1%     = -3.679
 Critical value 5%     = -2.968
 Critical value 10%    = -2.623
 => P-Value = 0.0. Rejecting Null Hypothesis.
 => Series is Stationary.


    Augmented Dickey-Fuller Test on "26" 
    -----------------------------------------------
 Null Hypothesis: Data has unit root. Non-Stationary.
 Significance Level    = 0.05
 Test Statistic        = -4.7452
 No. Lags Chosen       = 2
 Critical value 1%     = -3.7
 Critical value 5%     = -2.976
 Critical value 10%    = -2.628
 => P-Value = 0.0001. Rejecting Null Hypothesis.
 => Series is Stationary.


    Augmented Dickey-Fuller Test on "41" 
    -----------------------------------------------
 Null Hypothesis: Data has unit root. Non-Stationary.
 Significance Level    = 0.05
 Test

In [109]:
pivot1617 = pivot1617.diff().dropna()

In [110]:
for name, column in pivot1617.iteritems():
    adfuller_test(column, name=column.name)
    print('\n')

    Augmented Dickey-Fuller Test on "24" 
    -----------------------------------------------
 Null Hypothesis: Data has unit root. Non-Stationary.
 Significance Level    = 0.05
 Test Statistic        = -4.3199
 No. Lags Chosen       = 4
 Critical value 1%     = -3.738
 Critical value 5%     = -2.992
 Critical value 10%    = -2.636
 => P-Value = 0.0004. Rejecting Null Hypothesis.
 => Series is Stationary.


    Augmented Dickey-Fuller Test on "26" 
    -----------------------------------------------
 Null Hypothesis: Data has unit root. Non-Stationary.
 Significance Level    = 0.05
 Test Statistic        = -5.1009
 No. Lags Chosen       = 3
 Critical value 1%     = -3.724
 Critical value 5%     = -2.986
 Critical value 10%    = -2.633
 => P-Value = 0.0. Rejecting Null Hypothesis.
 => Series is Stationary.


    Augmented Dickey-Fuller Test on "41" 
    -----------------------------------------------
 Null Hypothesis: Data has unit root. Non-Stationary.
 Significance Level    = 0.05
 Te

# PIVOT 1776

In [111]:
for name, column in pivot1776.iteritems():
    adfuller_test(column, name=column.name)
    print('\n')

    Augmented Dickey-Fuller Test on "24" 
    -----------------------------------------------
 Null Hypothesis: Data has unit root. Non-Stationary.
 Significance Level    = 0.05
 Test Statistic        = -5.1202
 No. Lags Chosen       = 0
 Critical value 1%     = -3.679
 Critical value 5%     = -2.968
 Critical value 10%    = -2.623
 => P-Value = 0.0. Rejecting Null Hypothesis.
 => Series is Stationary.


    Augmented Dickey-Fuller Test on "26" 
    -----------------------------------------------
 Null Hypothesis: Data has unit root. Non-Stationary.
 Significance Level    = 0.05
 Test Statistic        = -2.9134
 No. Lags Chosen       = 5
 Critical value 1%     = -3.738
 Critical value 5%     = -2.992
 Critical value 10%    = -2.636
 => P-Value = 0.0438. Rejecting Null Hypothesis.
 => Series is Stationary.


    Augmented Dickey-Fuller Test on "41" 
    -----------------------------------------------
 Null Hypothesis: Data has unit root. Non-Stationary.
 Significance Level    = 0.05
 Te

In [112]:
pivot1776 = pivot1776.diff().dropna()

In [113]:
for name, column in pivot1776.iteritems():
    adfuller_test(column, name=column.name)
    print('\n')

    Augmented Dickey-Fuller Test on "24" 
    -----------------------------------------------
 Null Hypothesis: Data has unit root. Non-Stationary.
 Significance Level    = 0.05
 Test Statistic        = -5.8821
 No. Lags Chosen       = 1
 Critical value 1%     = -3.7
 Critical value 5%     = -2.976
 Critical value 10%    = -2.628
 => P-Value = 0.0. Rejecting Null Hypothesis.
 => Series is Stationary.


    Augmented Dickey-Fuller Test on "26" 
    -----------------------------------------------
 Null Hypothesis: Data has unit root. Non-Stationary.
 Significance Level    = 0.05
 Test Statistic        = -3.3624
 No. Lags Chosen       = 5
 Critical value 1%     = -3.753
 Critical value 5%     = -2.998
 Critical value 10%    = -2.639
 => P-Value = 0.0123. Rejecting Null Hypothesis.
 => Series is Stationary.


    Augmented Dickey-Fuller Test on "41" 
    -----------------------------------------------
 Null Hypothesis: Data has unit root. Non-Stationary.
 Significance Level    = 0.05
 Test

# PIVOT 624

removing seller 188 because it has too many zero values

In [114]:
pivot624.drop(188, axis=1, inplace=True)

In [115]:
for name, column in pivot624.iteritems():
    adfuller_test(column, name=column.name)
    print('\n')

    Augmented Dickey-Fuller Test on "23" 
    -----------------------------------------------
 Null Hypothesis: Data has unit root. Non-Stationary.
 Significance Level    = 0.05
 Test Statistic        = -4.6546
 No. Lags Chosen       = 0
 Critical value 1%     = -3.679
 Critical value 5%     = -2.968
 Critical value 10%    = -2.623
 => P-Value = 0.0001. Rejecting Null Hypothesis.
 => Series is Stationary.


    Augmented Dickey-Fuller Test on "24" 
    -----------------------------------------------
 Null Hypothesis: Data has unit root. Non-Stationary.
 Significance Level    = 0.05
 Test Statistic        = -5.3696
 No. Lags Chosen       = 0
 Critical value 1%     = -3.679
 Critical value 5%     = -2.968
 Critical value 10%    = -2.623
 => P-Value = 0.0. Rejecting Null Hypothesis.
 => Series is Stationary.


    Augmented Dickey-Fuller Test on "26" 
    -----------------------------------------------
 Null Hypothesis: Data has unit root. Non-Stationary.
 Significance Level    = 0.05
 Te

In [116]:
pivot624 = pivot624.diff().dropna()

In [117]:
for name, column in pivot624.iteritems():
    adfuller_test(column, name=column.name)
    print('\n')

    Augmented Dickey-Fuller Test on "23" 
    -----------------------------------------------
 Null Hypothesis: Data has unit root. Non-Stationary.
 Significance Level    = 0.05
 Test Statistic        = -7.7691
 No. Lags Chosen       = 0
 Critical value 1%     = -3.689
 Critical value 5%     = -2.972
 Critical value 10%    = -2.625
 => P-Value = 0.0. Rejecting Null Hypothesis.
 => Series is Stationary.


    Augmented Dickey-Fuller Test on "24" 
    -----------------------------------------------
 Null Hypothesis: Data has unit root. Non-Stationary.
 Significance Level    = 0.05
 Test Statistic        = -6.115
 No. Lags Chosen       = 1
 Critical value 1%     = -3.7
 Critical value 5%     = -2.976
 Critical value 10%    = -2.628
 => P-Value = 0.0. Rejecting Null Hypothesis.
 => Series is Stationary.


    Augmented Dickey-Fuller Test on "26" 
    -----------------------------------------------
 Null Hypothesis: Data has unit root. Non-Stationary.
 Significance Level    = 0.05
 Test Sta

The aim is to verify the presence of automated price systems and the presence of leaders and followers for November as well. The method used in the quarter analysis requires stationarity for all time series. 
We therefore tested this option with AD Fuller Test, and not all time series were stationary. So we tried a different approach for November, with the aim of identifying leaders and followers. 

In [119]:
df=novembre.copy()

#absolute value of the price change
df['price_change'] = df['price_change'].abs()

##fill na with 0
df['price_change'] = df['price_change'].fillna(0) 
df["yes_no"] = np.where(df["price_change"] > 0, 1, 0)

#filter only for yes_no = 1 (yes means that there is a price change)
df = df[df['yes_no'] == 1]

df.sort_values(['comp_date'], inplace=True)

For each product, the difference (in terms of days) of each price change made by a seller was calculated

In [120]:
df["comp_date"] = pd.to_datetime(df["comp_date"])
df["days_diff"]=df.groupby(['product_id'])["comp_date"].diff().dt.days

In [121]:
#calculate the mean of the days_diff for each seller and product
df2=df.copy()
df2=df2.groupby(['product_id','seller_id'])['days_diff'].mean()

#discard the rows with nan values
df2=df2.dropna()

#convert the series to dataframe
df2=pd.DataFrame(df2)

The mean of the difference in terms of days when the seller changes the price of a speficic product 

In [122]:
df2.head()

days_diff
product_id seller_id           
100043     24             3.000
           26             3.000
           48            10.000
100044     24             3.000
           26             3.000

In [123]:
#show df2 without base 10 exponential
pd.set_option('display.float_format', lambda x: '%.3f' % x)

In [124]:
#for each product_id sort the sellers by the mean of the days_diff
df2=df2.groupby(['product_id']).apply(lambda x: x.sort_values(['days_diff'], ascending=True))
df2

days_diff
product_id product_id seller_id           
100043     100043     24             3.000
                      26             3.000
                      48            10.000
100044     100044     24             3.000
                      26             3.000
...                                    ...
183889     183889     26             0.000
                      48             0.000
                      24             3.000
184287     184287     24             0.000
                      26             1.000

[12943 rows x 1 columns]

In [125]:
#for each product_id calculate the difference betweene each seller
df2['diff'] = df2.groupby(['product_id'])['days_diff'].diff()

The 'diff' column shows the distance between the price changes of each seller. for example for product 100043, sellers 24 and 26 change their price on average every three days, so their price change frequency is similar. This may indicate a synchronicity in setting prices, thus the presentation of dynamic pricing. 

In [126]:
df2.head(10)

days_diff  diff
product_id product_id seller_id                 
100043     100043     24             3.000   NaN
                      26             3.000 0.000
                      48            10.000 7.000
100044     100044     24             3.000   NaN
                      26             3.000 0.000
                      48            10.000 7.000
100052     100052     26             0.000   NaN
                      24             1.000 1.000
100078     100078     24             3.000   NaN
                      188            4.000 1.000

# Leaders and Followers Analysis

In [127]:
times_change=df.groupby(["product_id", "seller_id","comp_date"])["yes_no"].sum()
times_change=pd.DataFrame(times_change)

#remove levels
times_change=times_change.reset_index(level=[0,1,2])

'yes_no' column means that the product has undergone price changes 

In [128]:
times_change

product_id  seller_id  comp_date  yes_no
0          100043         24 2021-11-15       1
1          100043         24 2021-11-19       1
2          100043         26 2021-11-10       1
3          100043         26 2021-11-13       1
4          100043         48 2021-11-29       1
...           ...        ...        ...     ...
40166      183889         26 2021-11-16       1
40167      183889         48 2021-11-16       1
40168      184287         24 2021-11-28       1
40169      184287         26 2021-11-29       1
40170      184287         48 2021-11-28       1

[40171 rows x 4 columns]

In [129]:
#count times the price changed for each product_id and seller_id
count_change =times_change.groupby(["product_id", "seller_id"])["yes_no"].sum()

In [130]:
#count_change to dataframe
count_change=pd.DataFrame(count_change)
count_change

yes_no
product_id seller_id        
100043     24              2
           26              2
           48              1
100044     24              2
           26              2
...                      ...
183889     26              1
           48              1
184287     24              1
           26              1
           48              1

[14003 rows x 1 columns]

In [131]:
#map value yes_no for each product_id and seller_id to times_change and create new column
#the column represents the amount of times that the product changes the price
times_change['amount_change']=times_change.set_index(['product_id', 'seller_id']).index.map(count_change['yes_no'])


In [132]:
#drop times_change
#times_change.drop("times_change", axis=1, inplace=True)

In [133]:
#remove duplicate where product_id seller_id and amount_change are the same
#or each seller there are various repetitions in the dates, we took the least recent date (keep = 'first) because it indicates the first change price
times_change.drop_duplicates(subset=['product_id', 'seller_id', 'amount_change'],inplace=True, keep='first')
times_change

product_id  seller_id  comp_date  yes_no  amount_change
0          100043         24 2021-11-15       1              2
2          100043         26 2021-11-10       1              2
4          100043         48 2021-11-29       1              1
5          100044         24 2021-11-15       1              2
7          100044         26 2021-11-10       1              2
...           ...        ...        ...     ...            ...
40166      183889         26 2021-11-16       1              1
40167      183889         48 2021-11-16       1              1
40168      184287         24 2021-11-28       1              1
40169      184287         26 2021-11-29       1              1
40170      184287         48 2021-11-28       1              1

[14003 rows x 5 columns]

We have observed that, if automated pricing between two sellers is assumed, they will be inclined to follow the same trend; one imitates the other. So the number of times they change price is the same, which is why we only kept sellers who have an equal number of times they changed price.

In [134]:
#remove row where amount_change is different for each product and seller_id
times_change=times_change[times_change.duplicated(subset=['product_id', 'amount_change'], keep=False)]
times_change.head(10)

product_id  seller_id  comp_date  yes_no  amount_change
0       100043         24 2021-11-15       1              2
2       100043         26 2021-11-10       1              2
5       100044         24 2021-11-15       1              2
7       100044         26 2021-11-10       1              2
13      100078         24 2021-11-15       1              2
15      100078         26 2021-11-10       1              2
18      100078        188 2021-11-02       1              2
20      100083         24 2021-11-15       1              2
22      100083         26 2021-11-10       1              2
27      100204         24 2021-11-15       1              2

In [135]:
#for each product_id extract date that comes earlier which is the Leader date, and the date that comes later which is the Follower date
times_change['Leader_date'] = times_change.groupby(['product_id'])['comp_date'].transform('min')
times_change['Follower_date'] = times_change.groupby(['product_id'])['comp_date'].transform('max')


#if the Leader date is the same as the comp_date then the seller is the Leader
times_change['Lead/Follow'] = np.where(times_change['Leader_date'] == times_change['comp_date'], "Leader", "Follower")

In [136]:
#for each seller_id count the number of times that the seller is the Leader
times_change['Leader_count'] = times_change.groupby(['seller_id'])['Lead/Follow'].transform('count')

In [137]:
#number of times that each seller is the Leader
Leader_count=times_change.groupby(['seller_id'])['Leader_count'].max()
Leader_count=pd.DataFrame(Leader_count)
Leader_count

Leader_count
seller_id              
23                   24
24                 1998
26                 2279
41                  480
48                 1353
188                 581
407                 598
490                 363

In [138]:
#sort Leader_count by the number of times that each seller is the Leader
Leader_count=Leader_count.sort_values(['Leader_count'], ascending=False)

In [139]:
Leader_count

Leader_count
seller_id              
26                 2279
24                 1998
48                 1353
407                 598
188                 581
41                  480
490                 363
23                   24

In [140]:
#print total of Leader_count
Leader_count['Leader_count'].sum()

7676

In [141]:
#proportion of Leader_count over total for each seller
Leader_count['Leader_count_prop']=Leader_count['Leader_count']/Leader_count['Leader_count'].sum()
Leader_count

Leader_count  Leader_count_prop
seller_id                                 
26                 2279              0.297
24                 1998              0.260
48                 1353              0.176
407                 598              0.078
188                 581              0.076
41                  480              0.063
490                 363              0.047
23                   24              0.003